lxd/vm_qemu.go

```diff
@@ -0,0 +1,2329 @@
+		vendorData += `
+ - "cp /media/lxd_config/lxd-agent.service /etc/systemd/system/"
+ - "systemctl enable lxd-agent.service"
+ - "systemctl start lxd-agent"`
+	}
+
+	err = ioutil.WriteFile(configDrivePath+"/vendor-data", []byte(vendorData), 0600)
+	if err != nil {
+		return "", err
+	}
+
+	userData := vm.expandedConfig["user.user-data"]
+
+	// Use an empty user-data file if no custom user-data supplied.
+	if userData == "" {
+		userData = "#cloud-config"
+	}
+
+	err = ioutil.WriteFile(configDrivePath+"/user-data", []byte(userData), 0600)
+	if err != nil {
+		return "", err
+	}
+
+	metaData := fmt.Sprintf(`instance-id: %s
+local-hostname: %s
+`, vm.Name(), vm.Name())
+
+	err = ioutil.WriteFile(configDrivePath+"/meta-data", []byte(metaData), 0600)
+	if err != nil {
+		return "", err
+	}
+
+	lxdAgentServiceUnit := `[Unit]
+Description=LXD - agent
+After=media-lxd_config.mount
+[Service]
+Type=simple
+ExecStart=/media/lxd_config/lxd-agent
+[Install]
+WantedBy=multi-user.target
+`
+
+	err = ioutil.WriteFile(configDrivePath+"/lxd-agent.service", []byte(lxdAgentServiceUnit), 0600)
+	if err != nil {
+		return "", err
+	}
+
+	lxdConfigDriveMountUnit := `[Unit]
+Description = LXD - config drive
+Before=local-fs.target
+[Mount]
+Where=/media/lxd_config
+What=/dev/disk/by-label/cidata
+Type=iso9660
+[Install]
+WantedBy=multi-user.target
+`
+
+	err = ioutil.WriteFile(configDrivePath+"/media-lxd_config.mount", []byte(lxdConfigDriveMountUnit), 0600)
+	if err != nil {
+		return "", err
+	}
+
+	// Finally convert the config drive dir into an ISO file. The cidata label is important
+	// as this is what cloud-init uses to detect, mount the drive and run the cloud-init
+	// templates on first boot. The vendor-data template then modifies the system so that the
+	// config drive is mounted and the agent is started on subsequent boots.
+	isoPath := vm.Path() + "/config.iso"
+	_, err = shared.RunCommand("mkisofs", "-R", "-V", "cidata", "-o", isoPath, configDrivePath)
+	if err != nil {
+		return "", err
+	}
+
+	return isoPath, nil
+}
+
+// generateQemuConfigFile writes the qemu config file.
+func (vm *vmQemu) generateQemuConfigFile(configISOPath string, tapDev map[string]string) (string, error) {
+	pool, err := storagePools.GetPoolByInstance(vm.state, vm)
+	if err != nil {
+		logger.Warnf("Failed to init storage: %v", err)
+	}
+
+	_, rootDrive, err := pool.GetInstanceDisk(vm)
+	if err != nil {
+		logger.Warnf("Failed to get root disk: %v", err)
+	}
+
+	monitorPath := vm.getMonitorPath()
+	nvramPath := vm.getNvramPath()
+	vsockID := vm.vsockID()
+
+	conf := fmt.Sprintf(`
+# Machine
+[machine]
+graphics = "off"
+type = "q35"
+accel = "kvm"
+usb = "off"
+graphics = "off"
+[global]
+driver = "ICH9-LPC"
+property = "disable_s3"
+value = "1"
+[global]
+driver = "ICH9-LPC"
+property = "disable_s4"
+value = "1"
+[boot-opts]
+strict = "on"
+# CPU
+[smp-opts]
+cpus = "4"
+sockets = "1"
+cores = "2"
+threads = "2"
+# Memory
+[memory]
+size = "2G"
```

Comment:
@stgraber how do you want the RAM specified in config, and what should the default value be?

Reply:
Similar to CPU, if we don't have a `limits.memory`, we should have it line up with the host.
That's obviously going to be a bit of an issue for any VM that actually attempts to use that much memory, but will be consistent with what we do for containers and we'll obviously recommend users set appropriate limits.

Reply:
@stgraber will qemu actually take that amount of memory at start though, so setting it to the host's memory will cause issues?

Reply:
It shouldn't, no. The only case where it allocates all the memory configured is if using hugepages.

Reply:
@stgraber have added support for limits.memory, and if not supplied uses 1GB default RAM, as setting to full-host memory prevents VM starting with allocation error.